cloud/filestore/libs/vfs_fuse/fs_impl_list.cpp

```diff
@@ -398,4 +415,123 @@ void TFileSystem::ReleaseDir(
     ReplyError(*callContext, {}, req, 0);
 }
 
+bool TFileSystem::ValidateDirectoryHandle(
+    TCallContext& callContext,
+    fuse_req_t req,
+    fuse_ino_t ino,
+    uint64_t fh)
+{
+    std::shared_ptr<TDirectoryHandle> handle;
+    with_lock (CacheLock) {
+        auto it = DirectoryHandles.find(fh);
+        if (it == DirectoryHandles.end()) {
+            ReplyError(
+                callContext,
+                ErrorInvalidHandle(fh),
+                req,
+                EBADF);
+            return false;
+        }
+
+        handle = it->second;
+    }
+
+    Y_ABORT_UNLESS(handle);
+
+    if (!CheckDirectoryHandle(req, ino, *handle, Log, __func__)) {
+        ReplyError(callContext, ErrorInvalidHandle(fh), req, EBADF);
+        return false;
+    }
+
+    return true;
+}
+
+void TFileSystem::FSyncDir(
+    TCallContextPtr callContext,
+    fuse_req_t req,
+    fuse_ino_t ino,
+    int datasync,
+    fuse_file_info* fi)
+{
+    STORAGE_DEBUG("FSyncDir #" << ino << " @" << (fi ? fi->fh : -1llu));
+
+    if (!ValidateNodeId(*callContext, req, ino)) {
+        return;
+    }
+
+    const auto reqId = callContext->RequestId;
+
+    NProto::TProfileLogRequestInfo requestInfo;
+    InitProfileLogRequestInfo(
+        requestInfo,
+        EFileStoreFuseRequest::FsyncDir,
+        Now());
+    InitNodeInfo(
+        requestInfo,
+        datasync,
+        TNodeId{fi ? ino : InvalidNodeId},
+        THandle{fi ? fi->fh : InvalidHandle});
+
+    std::function<void(const TFuture<NProto::TError>&)> callback =
+        [=, ptr = weak_from_this(), requestInfo = std::move(requestInfo)](
+            const auto& future) mutable
+    {
+        auto self = ptr.lock();
+        if (!self) {
+            return;
+        }
+
+        const auto& response = future.GetValue();
+
+        FinalizeProfileLogRequestInfo(
+            std::move(requestInfo),
+            Now(),
+            self->Config->GetFileSystemId(),
+            response,
+            self->ProfileLog);
+
+        if (self->CheckError(*callContext, req, response)) {
+            self->ReplyError(*callContext, response, req, 0);
+        }
+    };
+
+    if (fi) {
```

Comment:
в каком случае в fi может быть null?

Reply:
looks like in `fsyncdir` `fi` can't be null [link](https://github.com/ydb-platform/nbs/blob/80591297ff72391f8c7870e1b2353ec3ff7c8605/contrib/libs/virtiofsd/fuse_lowlevel.c#L1404)

in `fsync` there is case where it can be null [link](https://github.com/ydb-platform/nbs/blob/80591297ff72391f8c7870e1b2353ec3ff7c8605/contrib/libs/virtiofsd/fuse_lowlevel.c#L1288) due to this [commit](https://github.com/qemu/qemu/commit/1b209805f8159c3f4d89ddb9390a5f64887cebff)